observability/WFLY-18582_add-prometheus-endpoint-to-micrometer-extension.adoc

```diff
@@ -0,0 +1,119 @@
+---
+categories:
+- observability
+stability-level: community
+issue: https://github.com/wildfly/wildfly-proposals/issues/590
+feature-team:
+ developer: jasondlee
+ sme:
+  - jamezp
+ outside-perspective:
+  - kabir
+---
+= [Community] WFLY-18582 - Add prometheus endpoint to micrometer extension
+:author:            Jason Lee
+:email:             jasondlee@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Micrometer replaced Micrometer Metrics as the user-facing metrics library provided by WildFly. As part of that
+transition, the export behavior moved from a pull-based Prometheus-like format to push-based mechanism using the
+OpenTelemetry Protocol. Based on feedback from users who still desire to use the pull-based approach, this change will
+add support for the Prometheus Registry in addition to or in place of the OTLP Registry.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-18582[WFLY-18582]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-2234[EAP7-2234]
+* https://issues.redhat.com/browse/EAP7-2235[EAP7-2235]
+
+
+=== Affected Projects or Components
+
+Only the Micrometer subsystem and related tests will be affected
+
+=== Other Interested Projects
+
+- OpenShift has an interest in Prometheus metrics
+
+=== Relevant Installation Types
+
+* Traditional standalone server (unzipped or provisioned by Galleon)
+* OpenShift Source-to-Image (S2I)
+* Managed domain
+* Bootable jar
+
+== Requirements
+
+* Existing servers, if configured to use Micrometer (e.g., the `endpoint` property has been set), will be migrated to be
+an OTLP-based configuration. Since OTLP is the only current option, there is no ambiguity in this migration.
+* Existing servers with the extension enabled but with no endpoint set will not be changed. That is to say, Micrometer
+will continue to use a no-op registry.
+* Support for a new registry, `PrometheusRegistry`, will be added.
+** It can be added as the only registry, or it can be added _in addition to_ the OTLP Registry
+** The registry will expose two model attribute:
+*** `context`: the context (e.g., `/prometheus`) on which the metrics will be exposed
+*** `security-enabled`: whether to secure the context using the user and roles defined in the application server
+** If the user would like to use `/metrics`, the WildFly metrics extension must be disabled first to avoid a naming conflict.
+* Micrometer, using its OTLP-based push approach, should continue to work in manage domain mode, but if Prometheus is configured to be used on the host controller, the startup attempt of the HC should fail.
```

Comment:
This needs clarification.

I believe the situation is that an attempt to add the new resource in a domain mode profile will be rejected, during execution of the resource's add operation. (That would be the correct behavior -- reject the attempt to add an illegal config as soon as it happens, which is when the user tries to add it to the domain profile.)

From there I don't think it's necessary to explain what will happen as a consequence of the add op failing as that is just standard WF behavior that applies to any resource add. If the add op happens during boot, boot fails, otherwise the DC reports the failure to the management client (CLI/HAL) that attempted the add.

I think there should be a requirement about noting this in the text description of relevant add op (i.e. the text in LocalDescriptions.properties)

Reply:
@bstansberry  Text updated. If you're happy with the update, please resolve this conversation. thx.

Reply:
Thanks; LGTM.